### DISCUSSION TITLE
What is the status of gc in wasm now?

### DISCUSSION THREAD

**Discussion by william4code** [WebAssembly/gc#407] [Q&A]
Hi, I am new to wasm.
I am concerning about the status of gc in wasm.
I have a language with gc mechanism which is implemented in runtime.
I can transform the language to LLVM IR.
If I want to transform LLVM IR to wasm and leverage the gc mechanism of wasm, what extra work I need to do?
Does the current implementation of gc in wasm work now?

**Comment by rossberg** [comment]
The GC proposal is currently at phase 3 of the [proposal process](https://github.com/WebAssembly/meetings/blob/main/process/phases.md) and is expected to move to phase 4 around September. At that point, it is effectively standard, modulo administrative matters. The proposal is currently implemented in V8, which plans to ship it by the end of the year, and there is ongoing work in Firefox and Wasmtime as well. You can follow the [roadmap](https://webassembly.org/roadmap/) for implementation status.

I am not aware of any work in the LLVM toolchain, so I don't know what hurdles that would imply. You'd probably have to use your own codegen for the time being.

**Reply by kripken** [reply]
+1, and adding info on the LLVM side, there has been an effort to add some amount of WasmGC support there:

https://github.com/Igalia/ref-cpp

But that effort was focused on the needs of C++ (to allow cycle collection), and it is less active currently.

Overall, I am not sure if we can expect to get full LLVM support for WasmGC, as in, being able to use it for a full language implementation. That would probably mean integrating with [LLVM's general GC support](https://llvm.org/docs/GarbageCollection.html), but that is very low-level (e.g., marking stuff on the stack as a root; GC objects reside in linear memory and can alias things there, etc.), and it's not clear to me if it can work with WasmGC which is higher-level.

So I expect people to continue to use their own codegen as current WasmGC toolchains have been, at least for the foreseeable future. We can at least share some amount of optimization effort across toolchains  since the Binaryen optimizer can optimize [WasmGC to WasmGC](https://github.com/WebAssembly/binaryen/wiki/GC-Optimization-Guidebook). For example, the J2Wasm (Java => WasmGC) toolchain depends on Binaryen for most optimizations.

**Reply by jryans** [reply]
> Overall, I am not sure if we can expect to get full LLVM support for WasmGC, as in, being able to use it for a full language implementation. That would probably mean integrating with [LLVM's general GC support](https://llvm.org/docs/GarbageCollection.html), but that is very low-level (e.g., marking stuff on the stack as a root; GC objects reside in linear memory and can alias things there, etc.), and it's not clear to me if it can work with WasmGC which is higher-level.

I am also not sure that this LLVM GC support is actually actively used and maintained these days... I've heard of various mishaps and sharp edges from groups who attempted to use it recently.

**Reply by Guo-yyds** [reply]
As of 2023-10-18, wasmgc still has no support. Neither Chrome nor firefox versions of the roadmap have been officially released.

**Reply by rossberg** [reply]
The proposal has been promoted to stage 4 last month, which means it's essentially finished, up to some editorial work.

It is turned on by default in Chrome 119, which currently is in beta channel and will make it to the release channel in the next 6-8 weeks. It is turned in Firefox 120 as well, which should be released in a similar time frame. Incremental rollout of Google sheets on Wasm GC has begun recently, i.e, if you open that with the latest Chrome, there is a 1% chance it's using the GC extension already.

See also [here](https://webassembly.org/roadmap/) (which currently is a bit behind, since multiple memories, threads, and relaxed SIMD are also "standardised" as of last week).
